src/v/rpc/transport.cc

```diff
@@ -81,10 +81,10 @@ ss::future<> transport::connect(clock_type::duration connection_timeout) {
 
 ss::future<>
 transport::connect(rpc::clock_type::time_point connection_timeout) {
+    _correlation_idx = 0;
+    _last_seq = sequence_t{0};
+    _seq = sequence_t{0};
     return base_transport::connect(connection_timeout).then([this] {
-        _correlation_idx = 0;
```

Comment:
> This may lead to situation in which two fibers trying to execute transport method would reuse the same correlation id.

both versions look problematic: last caller forcefully resets the state independent of what other fibers are doing. so is the real problem that more than one fiber calls connect? that strikes me as problematic, but if we have set things up to allow that, then perhaps we should identify that situation here and immediately return a ready future rather than letting a second fiber proceed?

Reply:
connect is executed under the semaphore, the problem was that after `do_connect` returned `is_valid` return `true`. Then in `reconnect_transport::get_connected` we return `transport` reference outside of the semaphore.

Reply:
is this a case where `transport` is not enforcing the rules of its own interface and needs to? if I knew that `transport::connect` was called under a semaphore I may not hesitate in the future to move this initialization code around, but clearly it is very sensitive to how external fibers are using the interface?

Reply:
maybe skipping `connect` if transport is valid would be a way to enforce the correct behavior ?

Reply:
Adding my own notes here after an offline discussion with Michal.

The primary problem is that the state changes that result in `is_valid` to become true during a reconnect are not executed atomically with these three statements that reset the sequencing.

So one fiber could do get_connected (is_valid is false) -> reconnection -> semaphore -> transport.connect -> base_transport::connect

and during base_transport::connect the connection is reset in a way that causes is_valid to be true. this is the last thing that happens, and so it is unlikely a scheduling point exists between this state change and the running of the continuation attached which resets the sequence numbers.

however that scheduling event can occur. and if it does then another fiber calling get_connected will find is_valid to be true and return a reference to the connection. then the other continuation resumes and resets the sequencing state which may have already been modified by the fiber that found a valid connection.

the result is that warnings about mismatched correlation ids can occur and make a lot of noise in the log.

some possible approaches to deal with this are to have an up call / virtual method from the base transport which allows the rpc transport to reset its state in an atomic way.